src/amberscript/parser.cc

```diff
@@ -1807,6 +1809,146 @@ Result Parser::ParseRun() {
   return Result("invalid token in RUN command: " + token->AsString());
 }
 
+Result Parser::ParseDebug() {
+  // DEBUG extends a RUN with debugger test cases
+  auto res = ParseRun();
+  if (!res.IsSuccess()) {
+    return res;
+  }
+
+  auto dbg = MakeUnique<debug::Script>();
+  for (auto token = tokenizer_->NextToken();; token = tokenizer_->NextToken()) {
+    if (token->IsEOL())
+      continue;
+    if (token->IsEOS())
+      return Result("missing DEBUG END command");
+    if (token->IsIdentifier() && token->AsString() == "END")
+      break;
+
+    if (token->AsString() == "THREAD") {
+      auto res = ParseDebugThread(dbg.get());
+      if (!res.IsSuccess()) {
+        return res;
+      }
+    } else {
+      return Result("invalid token in DEBUG command: " + token->AsString());
+    }
+  }
+
+  command_list_.back()->SetDebugScript(std::move(dbg));
+
+  return Result();
+}
+
+Result Parser::ParseDebugThread(debug::Events* dbg) {
+  auto result = Result();
+  auto parseThread = [&](debug::Thread* thread) {
```

Comment:
That is quite a big step away from the record-and-replay design I have, as creating the `debug::ThreadScript` and passing in would require splitting the `debug::Thread` interface into the existing debugger interface, and one with a `now-replay-this-thread-I've-already-made` method.

FWIW, the original implementation had the `BreakOn` methods return a `debug::Thread`, instead of the callback, the the actual debugger implementation needs these threads to be driven concurrently, and I don't want to expose the hairy threading logic outside of the private debugger implementation.

I understand your concerns with lambda lifetimes here, but I do feel like the compact nature of the shared interface outweighs the additional code required to make the script callback-free. I'm happy to try and find a compromise though. 

Unrelated to what you asked, I've moved the bulk of the lambda into a `ParseDebugThreadBody` method as this function was a bit unruly to read. This also cleans up the `Result` returns quite a bit.

Reply:
I think I didn't explain right. I think you can make the change I want and the external impact is in `BreakOnComputeGlobalInvocation` and `BreakOnVertexIndex` to just remove the first two lines. So, in ParseDebugThread we'd create the `debug::Thread` object. Call your ParseDebugThreadBody and pass in the thread and then pass the `thread` to the `BreakOn*` methods.

Basically, just makes the parsing straight line instead of taking a jog through the lambda. Does that makes sense?

(if what I"m saying still doesn't fit with what you want, please ignore me as this is technically fine, just took some thinking to figure out where the parse happens and what state the tokenizer would be in afterwards and I like to avoid thinking, heh)

Reply:
I think I understand you. Things just start unravelling at the:

 > Call your ParseDebugThreadBody and pass in the thread and then pass the thread to the BreakOn* methods.

We can expose the `ThreadScript` object (which is currently PImpl) - that's fine - and we can record the thread objects on this before handing it over to the `debug::Events`.

However, the symmetry and interface reuse falls apart when we actually come to doing the debugger replay. Currently the debugger internally creates a `std::thread` to concurrently run the threads, and on this `std::thread`, it calls the `BreakOn` callback, so that the recorded thread ops happen sequentially and block until they're complete. Given the current way things are structured, I don't see a trivial way to keep the interface the same after changing the parser to push the thread script.  

With all that said, I think I might have a way to do what you're after - but it requires a bit of restructuring. I'm hoping to get this into the hands of Jaebaek this week, so do you mind if I experiment a bit and come back to you with a follow up change?

Reply:
Sounds good. Please file a bug and add a TODO which references is.

Reply:
It was actually easier to implement what I had in mind than I was expecting. Please see the new CL at the end of the stack.

This does use `std::shared_ptr` (frowned upon in google style), but given these `ThreadScripts` need to pass through `std::function` (which doesn't allow move-only captures), and potential reuse of the scripts, this seemed to pass the 'good reason' clause.